src/Sylius/Component/Core/Promotion/Checker/Rule/ContainsProductRuleChecker.php

```diff
@@ -34,14 +37,13 @@ public function isEligible(PromotionSubjectInterface $subject, array $configurat
 
         /* @var $item OrderItemInterface */
         foreach ($subject->getItems() as $item) {
-            if ($configuration['variant'] != $item->getVariant()->getId()) {
-                continue;
-            }
+            if ($configuration['product'] === $item->getVariant()->getProduct()->getCode()) {
```

Comment:
Since you typehint on Core's `OrderItemInterface`, you can use `$item->getProduct()` instead of `$item->getVariant()->getProduct()`.